### PR TITLE
Color picker update with team color presets

### PR DESCRIPTION
**Pull request by drogoganor**
Changes to the color picker and ColorValidator to support built-in team colors.

colorpresets

Preset colors are defined in mod.yaml as TeamColorPresets. These are based off the colors in the original games, with a couple of extras I selected myself.

Users can also save a selected color in the Custom Colors by holding LCtrl and clicking a swatch.

Bots added to skirmish or multiplayer games will use a preset color if there are any available. If all preset colors are used, a random color will be used instead.

The previous attempt at this was: https://github.com/OpenRA/OpenRA/pull/13072

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/ColorPaletteSwatchWidget.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/ColorPaletteSwatchWidget.cs]: Outdated header.

**Comment by MustaphaTR** [comment]
> These are based off the colors in the original games, with a couple of extras I selected myself.

I think we should refer the tests done by @pi99y on #11248. Which was meant to be differ from eachother and tilesets as much as possible (he had 15 color for each mod tho, not 10). Rather than sticking to original colors.

File: OpenRA.Mods.Common/Widgets/ColorPaletteSwatchWidget.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/ColorPaletteSwatchWidget.cs]: Can you please add the click handling to `ColorBlockWidget` instead of creating a new widget type?

File: OpenRA.Game/Manifest.cs
Comment (pchote) [inline OpenRA.Game/Manifest.cs]: Please define these on the `ColorValidator` instead of creating a new engine-level field.  `IGlobalModData` implementations can automatically parse information out of the mod.yaml: see `MapGrid` for a simple example.

File: OpenRA.Mods.Common/ColorValidator.cs
Comment (pchote) [inline OpenRA.Mods.Common/ColorValidator.cs]: It would be better to stick with `HSLColor`s from the start, otherwise you need to needlessly convert between types.

File: mods/common/chrome/color-picker.yaml
Comment (pchote) [inline mods/common/chrome/color-picker.yaml]: Having everything on the one panel looks awkward IMO.
Can you put these into a separate tab instead?  Here is an old mockup from when the color picker was originally being built showing how it could look:

htoxo14

Comment (drogoganor) [reply mods/common/chrome/color-picker.yaml]: Thanks for the feedback. I'll give this another try.

**Comment by drogoganor** [comment]
Thanks all, I'll give this another shot soon.

**Comment by drogoganor** [comment]
I have a new version of this in the works, here's what it looks like so far. The colors used are the ones recommended by pi99y.

colorpresets2

Any advice you have would be welcomed. I'm not so sure of how it looks. Is a Clear button even required for the custom palette? Should it prompt the user to ensure they want to delete their custom palette?

**Comment by pchote** [comment]
The way I would do it is to make the custom colors a rolling queue and remove the Clear button.  Then, instead of using Ctrl+click, I would have an explicit "Store" button below the "Random" button on the side.

**Comment by drogoganor** [comment]
I've changed things to be that way, with a Store button. I quite like it. I will need to do some more rework though. I also increased the size of the swatches, to have two rows of 8.

colorpresets3

**Comment by GraionDilach** [comment]
I wouldn't go requiring 16 preset colors, but format the code to allow it being set up with 8 colors alone as well.

As #11248 proves, it's really hard to set up 16 significantly different colors on all terrain types of a mod. Sure, there is a setup for RA1 in the ticket, but the same color shades might not work for Tiberian Sun or the thirdparty mods (KKND, D2, RA2 comes in mind) which probably only deliver 8 colors from their respective games.

**Comment by drogoganor** [comment]
Here's what it looks like with just 8, a random shortlist of colors:

colorpresets4

I've made the swatches including the label be dynamically generated so we can choose how many we want of each.

**Comment by drogoganor** [comment]
For reference, the colors included for all 4 mods. TS is the same colors for CNC as this wasn't included in pi99y's analysis: https://github.com/OpenRA/OpenRA/issues/11248

colorpresets5

**Comment by drogoganor** [comment]
Here's what I have so far. Swatch is now a template item which can be resized in the yaml instead of constructed by code.

colorpresets6

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]: This can simplify to `newSwatch = template.Clone()`

**Comment by drogoganor** [comment]
New version having resized the mixer:

colorpresets6

**Comment by pchote** [comment]
Looks great!

One last UI nit: Can you please make the Random and Store buttons a few px wider so that they have the same margin on the left as on the right?

**Comment by drogoganor** [comment]
Good one on the dialog2. This is what it looks like now:

colorpresets6

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs
Comment (drogoganor) [reply OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]: Is it alright if it looks like this afterward?

```
var newSwatch = (ColorBlockWidget) template.Clone();
newSwatch.IsVisible = () => true;
newSwatch.Id = nameString;
newSwatch.X = x.ToString();
newSwatch.Y = y.ToString();
newSwatch.Initialize(new WidgetArgs());
```

**Comment by drogoganor** [comment]
Changed swatch creation to use a clone of the template swatch. I had to set Visible to true and also IsVisible = () => true; to get it to work correctly. Also fixed a bug with swatch removal after moving to the new UI element types.

File: OpenRA.Mods.Common/ColorValidator.cs
Comment (pchote) [inline OpenRA.Mods.Common/ColorValidator.cs]: Likewise here.

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]: IMO it would feel much better ingame if this always inserted the color at the start (`list.Insert(0, hexColor)`), rolling all the older colors one step forward in the list.

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]: Style nit: loop indices should be declared as `var`

**Review by pchote — COMMENTED** [review]
Looking good!

I have one moderate structural request that should simplify the implementation a lot.  I expect to have some smaller comments once this has been done.

**Comment by drogoganor** [comment]
Sure thing, using HSLColor was a really good idea. I didn't know it would automatically serialize like that. It simplified things a bit.

I've added the latest change as a separate commit, because it might be a bit easier to review that way. Although if I should use a single squashed commit instead, let me know. I was going to squash it into a single commit at the end once it gets approved.

**Review by pchote — COMMENTED** [review]
Here are a few more suggestions to further simplify things:

**Comment by drogoganor** [comment]
I've added the suggested changes and merged into a single commit. Much neater code!

**Comment by drogoganor** [comment]
`var paletteCols = 8;
var palettePresetRows = 2;
var paletteCustomRows = 1;

MiniYaml yaml;
if (logicArgs.TryGetValue("PaletteColumns", out yaml))
	paletteCols = Convert.ToInt32(yaml.Value);
if (logicArgs.TryGetValue("PalettePresetRows", out yaml))
	palettePresetRows = Convert.ToInt32(yaml.Value);
if (logicArgs.TryGetValue("PaletteCustomRows", out yaml))
	paletteCustomRows = Convert.ToInt32(yaml.Value);`

Something like this or can it be a bit neater somehow?

**Review by pchote — COMMENTED** [review]
A few last polish nits, then LGTM!

Can you please also apply the following yaml changes?

```diff
diff --git a/mods/cnc/chrome/color-picker.yaml b/mods/cnc/chrome/color-picker.yaml
index 6507171228..e7419c6116 100644
--- a/mods/cnc/chrome/color-picker.yaml
+++ b/mods/cnc/chrome/color-picker.yaml
@@ -5,26 +5,26 @@ Background@COLOR_CHOOSER:
 		PaletteCustomRows: 1
 	Background: panel-black
 	Width: 311
-	Height: 150
+	Height: 148
 	Children:
 		Button@RANDOM_BUTTON:
 			Key: tab
 			X: 229
-			Y: 91
+			Y: 89
 			Width: 77
 			Height: 25
 			Text: Random
 		Button@STORE_BUTTON:
 			X: 229
-			Y: 120
+			Y: 118
 			Width: 77
 			Height: 25
 			Text: Store
 			Font: Bold
 		ActorPreview@PREVIEW:
-			X: 230
+			X: 232
 			Y: 7
-			Width: 80
+			Width: 77
 			Height: 73
 			Animate: true
 		Button@MIXER_TAB_BUTTON:
@@ -41,8 +41,7 @@ Background@COLOR_CHOOSER:
 			Width: 80
 			Text: Palette
 			Font: Bold
-		Background@MIXER_TAB:
-			Background: panel-black
+		Container@MIXER_TAB:
 			X: 5
 			Y: 5
 			Width: PARENT_RIGHT - 91
@@ -66,7 +65,7 @@ Background@COLOR_CHOOSER:
 					X: 0
 					Y: 22
 					Width: PARENT_RIGHT
-					Height: 94
+					Height: 92
 					Children:
 						ColorMixer@MIXER:
 							X: 2
@@ -104,7 +103,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 58
 							X: 2
-							Y: 15
+							Y: 16
 							Children:
 								ColorBlock@COLORPRESET:
 									X: 0
@@ -117,7 +116,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 13
 							X: 2
-							Y: 72
+							Y: 71
 							Children:
 								Label@LABEL:
 									Font: TinyBold
diff --git a/mods/common/chrome/color-picker.yaml b/mods/common/chrome/color-picker.yaml
index 81e4ef7ad1..d67611fca3 100644
--- a/mods/common/chrome/color-picker.yaml
+++ b/mods/common/chrome/color-picker.yaml
@@ -5,27 +5,27 @@ Background@COLOR_CHOOSER:
 		PaletteCustomRows: 1
 	Background: dialog2
 	Width: 326
-	Height: 156
+	Height: 154
 	Children:
 		Button@RANDOM_BUTTON:
 			Key: tab
 			X: 245
-			Y: 97
+			Y: 95
 			Width: 76
 			Height: 25
 			Text: Random
 			Font: Bold
 		Button@STORE_BUTTON:
 			X: 245
-			Y: 126
+			Y: 124
 			Width: 76
 			Height: 25
 			Text: Store
 			Font: Bold
 		ActorPreview@PREVIEW:
 			X: 245
-			Y: 21
-			Width: 80
+			Y: 13
+			Width: 76
 			Height: 73
 		Button@MIXER_TAB_BUTTON:
 			X: 5
@@ -41,8 +41,7 @@ Background@COLOR_CHOOSER:
 			Width: 80
 			Text: Palette
 			Font: Bold
-		Background@MIXER_TAB:
-			Background: dialog3
+		Container@MIXER_TAB:
 			X: 5
 			Y: 5
 			Width: PARENT_RIGHT - 90
@@ -104,7 +103,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 58
 							X: 2
-							Y: 15
+							Y: 16
 							Children:
 								ColorBlock@COLORPRESET:
 									X: 0
@@ -117,7 +116,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 13
 							X: 2
-							Y: 76
+							Y: 75
 							Children:
 								Label@LABEL:
 									Font: TinyBold
diff --git a/mods/ts/chrome/color-picker.yaml b/mods/ts/chrome/color-picker.yaml
index 93f9bc184f..b833c71469 100644
--- a/mods/ts/chrome/color-picker.yaml
+++ b/mods/ts/chrome/color-picker.yaml
@@ -5,27 +5,27 @@ Background@COLOR_CHOOSER:
 		PaletteCustomRows: 1
 	Background: dialog2
 	Width: 326
-	Height: 156
+	Height: 154
 	Children:
 		Button@RANDOM_BUTTON:
 			Key: tab
 			X: 245
-			Y: 97
+			Y: 95
 			Width: 76
 			Height: 25
 			Text: Random
 			Font: Bold
 		Button@STORE_BUTTON:
 			X: 245
-			Y: 126
+			Y: 124
 			Width: 76
 			Height: 25
 			Text: Store
 			Font: Bold
 		ActorPreview@PREVIEW:
 			X: 245
-			Y: 11
-			Width: 80
+			Y: 16
+			Width: 76
 			Height: 74
 			Animate: true
 		Button@MIXER_TAB_BUTTON:
@@ -42,8 +42,7 @@ Background@COLOR_CHOOSER:
 			Width: 80
 			Text: Palette
 			Font: Bold
-		Background@MIXER_TAB:
-			Background: dialog3
+		Container@MIXER_TAB:
 			X: 5
 			Y: 5
 			Width: PARENT_RIGHT - 90
@@ -105,7 +104,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 58
 							X: 2
-							Y: 15
+							Y: 16
 							Children:
 								ColorBlock@COLORPRESET:
 									X: 0
@@ -118,7 +117,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 13
 							X: 2
-							Y: 76
+							Y: 75
 							Children:
 								Label@LABEL:
 									Font: TinyBold
```

**Comment by drogoganor** [comment]
Thanks. I've updated the code and yaml. Also rebased to the latest bleed.

**Review by reaperrr — DISMISSED** [review]

**Comment by reaperrr** [comment]
Looks good to me, LGTM once @pchote has given this a last look-over.

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]: This isn't a scrollpanel, so can you please rename to `var palettePanel` and `PALETTE_TAB_PANEL`?

**Review by pchote — DISMISSED** [review]
One last minor nit, then LGTM

**Review by pchote — APPROVED** [review]

**Comment by drogoganor** [comment]
Thanks for all your help!

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/57fdded66d7e3182066d3c5ac3724279b0ff55c7

**Comment by pi99y** [comment]
Yay! Thank you guys!

**Pull request merged**
